File: backend/src/sustainability_desk/contract/stored_report_state.py
# ABOUTME: 轻量版 StoredReportState 的严格持久化合同，按版本定义报告实例快照允许保存的字段。
# ABOUTME: V4 删除全局缺失策略；必填值仍由阶段化 readiness 判定，不由存储 schema 冒充业务门禁。
# ABOUTME(en): Strict persistence contract for the lightweight StoredReportState, versioned per snapshot field set.
# ABOUTME(en): V4 drops the global missing-value policy; required values stay with phased readiness, not the schema.
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from sustainability_desk.contract.models import (
    AppendixPackage,
    BlockState,
    DisclosureProfile,
    Inline,
    MaterialityAssessmentInput,
    PrimaryInputMode,
    QuantitativeMetricsMeta,
    SectionDisplayTitle,
    StakeholderEngagementProfile,
)
from sustainability_desk.contract.structured_inputs import StructuredInputFreshness

STORED_REPORT_STATE_VERSION = 4


class StoredReportStateModel(BaseModel):
    """状态快照及其嵌套 DTO 的严格解析基类。"""

    model_config = ConfigDict(extra="forbid")


class StoredIntakeAnswer(StoredReportStateModel):
    """用户答案快照项：只保留实例答案，不接受题目结构。"""

    answer: str | list[str] | None = None
    supplement: str | None = None


class StoredCompanyBusinessSummary(StoredReportStateModel):
    """公司业务摘要快照项：派生值与其来源指纹同体保存。

    指纹取自压缩来源 `company_profile`；来源变化即摘要过期，须重算。
    值与指纹分开保存会让两者可能不一致，故合为一项。
    """

    text: str = Field(min_length=1)
    sourceFingerprint: str = Field(pattern=r"^[0-9a-f]{64}$")


class StoredGeneratedBlock(StoredReportStateModel):
    """生成段落快照项：仅允许正文与生命周期状态。"""

    content: list[Inline] | None = None
    state: BlockState | None = None


class StoredImageBlock(StoredReportStateModel):
    """素材图片承载位快照项：仅保存按序放置的资产引用，题注与替代文本归资产 owner。"""

    layoutAssetIds: list[UUID] = Field(
        default_factory=list,
        json_schema_extra={"default": []},
    )
    state: Literal["ready"] = "ready"


class StoredTableCell(StoredReportStateModel):
    """表格单元格快照：保留用户值，不接受 Plate 文本骨架或列合同。"""

    type: Literal["td", "th"] = "td"
    colKey: str | None = None
    value: str | list[str] | None = None
    options: list[str] | None = None
    colSpan: int = Field(default=1, ge=1)
    rowSpan: int = Field(default=1, ge=1)
    cellState: BlockState | None = None


class StoredRowOrigin(StoredReportStateModel):
    """AI 定行的可重生成事实；不包含行级生成规格。"""

    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
    )

    from_: Literal["ai", "user"] = Field(default="ai", alias="from")
    theme: str | None = None
    category: str | None = None
    driver_hint: str | None = None


class StoredTableRow(StoredReportStateModel):
    """表格用户行快照：结构、generation 与 appears_when 均由运行时模板拥有。"""

    type: Literal["tr"] = "tr"
    headerRow: bool = False
    state: BlockState | None = None
    origin: StoredRowOrigin | None = None
    children: list[StoredTableCell] = Field(
        default_factory=list,
        json_schema_extra={"default": []},
    )


class StoredTableBlock(StoredReportStateModel):
    """表格实例快照项：表头和列合同仍由运行时模板提供。

    ``state="omitted"`` 与空行共同表示本轮没有足以形成数据行的用户资料；
    该处置是报告状态的一部分，不能由导出器临时猜测。
    """

    children: list[StoredTableRow] = Field(
        default_factory=list,
        json_schema_extra={"default": []},
    )
    state: BlockState | None = None


class StoredReportMetaV4(StoredReportStateModel):
    """V4 报告级状态；缺失处置由输入义务与 Block 合同拥有。"""

    quantitativeMetrics: QuantitativeMetricsMeta = Field(
        default_factory=QuantitativeMetricsMeta
    )
    materialityStrategy: Literal["complete_coverage"] | None = None
    # 报告级填报方式二选一；None = 用户尚未选择。只驱动分步流编排与义务归属，
    # 不改写任何已填输入（两条证据轨在生成侧本就允许任一为空）。
    primaryInputMode: PrimaryInputMode | None = None


#: 服务端拥有的字段：只由专用写入器改写，generic 客户端通道只能原样往返。
#:
#: 这是所有权的**唯一声明处**——`put_state` 的锁内保全由它派生，不再在函数体里
#: 逐个枚举。默认方向是「受保护」：契约新增报告级字段时若未登记为前端拥有，
#: generic 通道就抹不掉它。若改为白名单枚举、新增字段默认不受保护，
#: 新提升为报告级事实的字段（如 `meta.primaryInputMode`）一旦漏配，
#: 客户端整体 PUT 就会把它写回 null，用户答完必答题仍被生成门禁挡住。
#:
#: 顶层字段用字段名；`meta` 的子字段用 `meta.` 前缀。
SERVER_OWNED_STATE_FIELDS: frozenset[str] = frozenset(
    {
        "assessmentInput",
        "structuredInputFreshness",
        # 派生值，由生成编排写入并自带来源指纹；用户没有「清空业务摘要」这个动作。
        "companyBusinessSummary",
        "meta.materialityStrategy",
        "meta.quantitativeMetrics",
        # 编排事实，只由 PATCH /primary-input-mode 写；None = 尚未选择，
        # 产品上没有「取消填报方式」这个动作，因此保全不损失任何功能。
        "meta.primaryInputMode",
    }
)


class StoredReportStateV4(StoredReportStateModel):
    """V4 严格状态边界；不接受旧缺失策略，未完成输入交由 readiness 处理。"""

    version: Literal[4]
    fields: dict[str, str | int | float | None] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    intakeItems: dict[str, StoredIntakeAnswer] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    assessmentInput: MaterialityAssessmentInput | None = None
    disclosureProfile: DisclosureProfile | None = None
    appendixPackage: AppendixPackage | None = None
    meta: StoredReportMetaV4 | None = None
    stakeholderEngagement: StakeholderEngagementProfile | None = None
    sectionTitles: dict[str, SectionDisplayTitle] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    generatedBlocks: dict[str, StoredGeneratedBlock] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    tableBlocks: dict[str, StoredTableBlock] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    imageBlocks: dict[str, StoredImageBlock] = Field(
        default_factory=dict,
        json_schema_extra={"default": {}},
    )
    structuredInputFreshness: StructuredInputFreshness = Field(
        default_factory=StructuredInputFreshness
    )
    # 派生值不进 fields（fields 只承载用户输入）；派生态与其来源指纹在此单独持有。
    companyBusinessSummary: StoredCompanyBusinessSummary | None = None

    @model_validator(mode="before")
    @classmethod
    def _reject_retired_version(cls, value: object) -> object:
        """旧本地报告不做迁移或双读，明确要求用户重建。"""

        if (
            isinstance(value, dict)
            and "version" in value
            and value.get("version") != STORED_REPORT_STATE_VERSION
        ):
            raise ValueError("状态版本已废弃，请重建报告")
        return value


def empty_stored_report_state() -> StoredReportStateV4:
    """构造新报告的唯一空 V4 状态。"""

    return StoredReportStateV4(version=STORED_REPORT_STATE_VERSION)

File: frontend/lib/stored-report-state.ts
// ABOUTME: 浏览器与服务端报告状态快照的 parse-first 边界，只保存用户事实和实例内容。
// ABOUTME: StoredReportStateV4 类型与 Ajv schema 均由后端独立领域合同生成；旧本地快照由应用边界明确废弃。
import { parseStoredReportStateValue } from "./contract-parser";
import type {
  StoredReportStateV4,
  StoredTableRow,
} from "./stored-report-state.generated";
import type { Block, Report } from "./schema";

export type { StoredReportStateV4 } from "./stored-report-state.generated";

export const STORED_REPORT_STATE_VERSION = 4;

export class StoredReportStateVersionError extends Error {
  constructor(readonly actualVersion: unknown) {
    super(`不支持的报告状态版本：${String(actualVersion)}`);
    this.name = "StoredReportStateVersionError";
  }
}

export class StoredReportStateJSONError extends Error {}

function emptyStoredReportState(): StoredReportStateV4 {
  return {
    version: STORED_REPORT_STATE_VERSION,
    fields: {},
    intakeItems: {},
    assessmentInput: null,
    disclosureProfile: null,
    appendixPackage: null,
    meta: null,
    stakeholderEngagement: null,
    sectionTitles: {},
    generatedBlocks: {},
    tableBlocks: {},
    imageBlocks: {},
    companyBusinessSummary: null,
  };
}

function walkBlocks(sections: Report["sections"], visit: (block: Block) => void) {
  for (const section of sections ?? []) {
    for (const block of section.blocks ?? []) visit(block);
    walkBlocks(section.children ?? [], visit);
  }
}

function walkSections(
  sections: Report["sections"],
  visit: (section: Report["sections"][number]) => void,
) {
  for (const section of sections ?? []) {
    visit(section);
    walkSections(section.children ?? [], visit);
  }
}

function isGeneratedParagraph(block: Block): boolean {
  return block.type === "paragraph" && (
    block.blockType === "generative" || block.blockType === "constrained"
  );
}

function isPersistedTableBlock(block: Block): boolean {
  return block.type === "table"
    && !!block.table
    && block.source !== "derived"
    && block.table.rowSource !== "stakeholder_engagement";
}

function fieldValues(report: Report): StoredReportStateV4["fields"] {
  const values: StoredReportStateV4["fields"] = {};
  for (const [key, field] of Object.entries(report.fields ?? {})) {
    values[key] = field.value ?? null;
  }
  return values;
}

function intakeAnswers(report: Report): StoredReportStateV4["intakeItems"] {
  const answers: StoredReportStateV4["intakeItems"] = {};
  for (const item of report.intakeItems ?? []) {
    if (item.answer !== undefined || item.supplement !== undefined) {
      answers[item.key] = {
        answer: item.answer ?? null,
        supplement: item.supplement ?? null,
      };
    }
  }
  return answers;
}

function storedTableRows(rows: NonNullable<Block["table"]>["children"]): StoredTableRow[] {
  return (rows ?? []).map((row) => ({
    type: row.type ?? "tr",
    headerRow: row.headerRow ?? false,
    state: row.state ?? null,
    origin: row.origin
      ? {
          from: row.origin.from ?? "ai",
          theme: row.origin.theme ?? null,
          category: row.origin.category ?? null,
          driver_hint: row.origin.driver_hint ?? null,
        }
      : null,
    children: (row.children ?? []).map((cell) => ({
      type: cell.type ?? "td",
      colKey: cell.colKey ?? null,
      value: cell.value ?? null,
      options: cell.options ?? null,
      colSpan: cell.colSpan ?? 1,
      rowSpan: cell.rowSpan ?? 1,
      cellState: cell.cellState ?? null,
    })),
  }));
}

/**
 * 新建报告的首个状态：只承载新报告真实拥有的事实——输入字段默认值与模板初始配置。
 *
 * 不走 reportToStoredReportState：那个函数按模板装配重建全量投影，会把尚未生成的
 * 空块骨架（content/state 均为 null 的 generatedBlocks、只有表头行的 tableBlocks）
 * 一并写入。这些骨架不是用户事实；写入边界按报告侧口径全等比较，
 * 会把范围外章节的骨架判为范围外而 403，使新建流程在第二步失败、
 * 报告建成却无法进入。
 */
export function newReportStoredState(report: Report): StoredReportStateV4 {
  const state = emptyStoredReportState();
  state.fields = fieldValues(report);
  state.disclosureProfile = report.disclosureProfile ?? null;
  state.appendixPackage = report.appendixPackage ?? null;
  return state;
}

/** 状态侧 meta → 报告内容 meta：剔除只属于编排的字段。
 *
 * `primaryInputMode`（填报方式）是报告级**编排**事实，落在状态 meta 上供生成闸与导出闸
 * 共同裁定；但 Report 合同 `extra="forbid"`，把它原样带进 Report.meta 会让 /api/plan
 * 整体 422、编制页只剩一行错误。投影必须显式列出报告内容字段，
 * 不做整体透传——下次再往状态 meta 加编排字段时，这里的遗漏才不会变成线上故障。
 */
function reportMetaFromStored(meta: StoredReportStateV4["meta"]): Report["meta"] | null {
  if (!meta) return null;
  return {
    quantitativeMetrics: meta.quantitativeMetrics,
    materialityStrategy: meta.materialityStrategy ?? null,
  } as Report["meta"];
}

export function reportToStoredReportState(report: Report): StoredReportStateV4 {
  const state = emptyStoredReportState();
  state.fields = fieldValues(report);
  state.intakeItems = intakeAnswers(report);
  state.assessmentInput = report.assessmentInput ?? null;
  state.disclosureProfile = report.disclosureProfile ?? null;
  state.appendixPackage = report.appendixPackage ?? null;
  state.meta = report.meta ?? null;
  state.stakeholderEngagement = report.stakeholderEngagement ?? null;
  walkSections(report.sections ?? [], (section) => {
    if (section.displayTitle) state.sectionTitles[section.key] = section.displayTitle;
  });
  walkBlocks(report.sections ?? [], (block) => {
    if (isGeneratedParagraph(block)) {
      state.generatedBlocks[block.id] = {
        content: block.content ?? null,
        state: block.state ?? null,
      };
    } else if (isPersistedTableBlock(block) && block.table) {
      state.tableBlocks[block.id] = {
        children: storedTableRows(block.table.children),
        state: block.state ?? null,
      };
    } else if (
      block.image?.layoutAssetSlot
      && (block.image.layoutAssetIds?.length ?? 0) > 0
    ) {
      // 素材承载位的放置结果必须随状态往返,否则工作台保存会清空已放置图片。
      state.imageBlocks[block.id] = {
        layoutAssetIds: block.image.layoutAssetIds ?? [],
        state: "ready",
      };
    }
  });
  // Validation applies schema defaults in place (`useDefaults`), and the snapshot shares
  // `meta` and other sub-objects with the runtime Report by reference. Validate a copy so
  // orchestration defaults such as `meta.primaryInputMode` never land on Report.meta, whose
  // server contract forbids extra fields (plan, diagnose and export would all reject it).
  return parseStoredReportStateValue(structuredClone(state));
}

export function parseStoredReportState(raw: string | null): StoredReportStateV4 {
  if (raw === null) return emptyStoredReportState();
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    throw new StoredReportStateJSONError("报告状态快照不是有效 JSON", { cause: error });
  }
  if (
    parsed === null
    || typeof parsed !== "object"
    || Array.isArray(parsed)
    || !("version" in parsed)
    || parsed.version !== STORED_REPORT_STATE_VERSION
  ) {
    const version = parsed !== null
      && typeof parsed === "object"
      && !Array.isArray(parsed)
      && "version" in parsed
      ? parsed.version
      : "缺失";
    throw new StoredReportStateVersionError(version);
  }
  return parseStoredReportStateValue(parsed);
}

export function parseLocalStoredReportState(raw: string | null): {
  state: StoredReportStateV4 | null;
  discardedObsoleteVersion: boolean;
} {
  try {
    return { state: parseStoredReportState(raw), discardedObsoleteVersion: false };
  } catch (error) {
    if (error instanceof StoredReportStateVersionError) {
      return { state: null, discardedObsoleteVersion: true };
    }
    throw error;
  }
}

function applyFieldValues(
  template: Report,
  state: StoredReportStateV4,
): Report["fields"] {
  const fields: Report["fields"] = {};
  for (const [key, field] of Object.entries(template.fields ?? {})) {
    fields[key] = {
      ...field,
      value: Object.prototype.hasOwnProperty.call(state.fields, key)
        ? state.fields[key]
        : field.value,
    };
  }
  return fields;
}

function applyIntakeAnswers(
  template: Report,
  state: StoredReportStateV4,
): Report["intakeItems"] {
  return (template.intakeItems ?? []).map((item) => {
    const stored = state.intakeItems[item.key];
    return stored
      ? {
          ...item,
          answer: stored.answer ?? null,
          supplement: stored.supplement ?? null,
        }
      : item;
  });
}

function applyStoredBlocks(
  sections: Report["sections"],
  state: StoredReportStateV4,
): Report["sections"] {
  return (sections ?? []).map((section) => ({
    ...section,
    displayTitle: state.sectionTitles[section.key] ?? section.displayTitle,
    blocks: (section.blocks ?? []).map((block) => {
      if (isGeneratedParagraph(block)) {
        const stored = state.generatedBlocks[block.id];
        return stored
          ? {
              ...block,
              content: stored.content ?? block.content,
              state: stored.state ?? block.state,
            }
          : block;
      }
      if (isPersistedTableBlock(block) && block.table) {
        const stored = state.tableBlocks[block.id];
        return stored
          ? {
              ...block,
              state: stored.state ?? block.state,
              table: {
                ...block.table,
                children: stored.children ?? block.table.children,
              },
            }
          : block;
      }
      if (block.image?.layoutAssetSlot) {
        const stored = state.imageBlocks[block.id];
        return stored
          ? {
              ...block,
              state: stored.state,
              image: { ...block.image, layoutAssetIds: stored.layoutAssetIds },
            }
          : block;
      }
      return block;
    }),
    children: section.children
      ? applyStoredBlocks(section.children, state)
      : section.children,
  }));
}

export function applyStoredStateToTemplate(
  template: Report,
  state: StoredReportStateV4,
): Report {
  const storedAppendix = state.appendixPackage;
  const templateAppendix = template.appendixPackage;
  return {
    ...template,
    fields: applyFieldValues(template, state),
    intakeItems: applyIntakeAnswers(template, state),
    assessmentInput: state.assessmentInput ?? template.assessmentInput ?? null,
    assessment: null,
    disclosureProfile: state.disclosureProfile ?? template.disclosureProfile ?? null,
    appendixPackage: storedAppendix
      ? {
          ...templateAppendix,
          ...storedAppendix,
          ...(storedAppendix.externalAssuranceReport
            ? {
                externalAssuranceReport: {
                  ...templateAppendix?.externalAssuranceReport,
                  ...storedAppendix.externalAssuranceReport,
                },
              }
            : {}),
          ...(storedAppendix.readerFeedbackContactInformation
            ? {
                readerFeedbackContactInformation: {
                  ...templateAppendix?.readerFeedbackContactInformation,
                  ...storedAppendix.readerFeedbackContactInformation,
                },
              }
            : {}),
        }
      : templateAppendix,
    meta: reportMetaFromStored(state.meta) ?? template.meta ?? null,
    stakeholderEngagement: state.stakeholderEngagement
      ?? template.stakeholderEngagement
      ?? null,
    sections: applyStoredBlocks(template.sections ?? [], state),
  };
}

/**
 * 以最近一次权威服务端 state 为基线,保全前端投影覆盖不到的内容。
 *
 * 浏览器运行时 Report 只含当前已装配的章节与答案键集;议题章节的生成正文、表格行、素材图片放置
 * 与 intake 答案(generatedBlocks/tableBlocks/imageBlocks/intakeItems 中不属于当前投影键集的键)
 * 在 reportToStoredReportState 的重建中会随装配变化而缺席。若直接整体 PUT,会把这些服务端事实
 * 无声抹掉——前端是窄投影的 owner,不是全量 state 的 owner。
 */
export function mergeStoredStatePreservingUnprojected(
  projected: StoredReportStateV4,
  baseline: StoredReportStateV4,
  projectedBlockIds: ReadonlySet<string>,
  projectedSectionKeys: ReadonlySet<string>,
  projectedIntakeItemKeys: ReadonlySet<string>,
): StoredReportStateV4 {
  const mergeByKeys = <T>(
    own: Record<string, T>,
    base: Record<string, T>,
    projectedKeys: ReadonlySet<string>,
  ): Record<string, T> => {
    const next: Record<string, T> = { ...own };
    for (const [key, value] of Object.entries(base)) {
      if (!projectedKeys.has(key) && !(key in next)) next[key] = value;
    }
    return next;
  };
  return {
    ...projected,
    // 以下两个字段是**服务端拥有**的（后端 SERVER_OWNED_STATE_FIELDS）：权威保全在
    // put_state 的锁内、与 CAS 同事务，锁内旧值定义上不可能陈旧。这里的回填只是
    // 第二道防线，用于减少无谓的字段抖动——它依赖「baseline 是新鲜的」这个前端
    // 无法自证的前提，因此不得被当作正确性来源（基线陈旧时这里的
    // `?? null` 恰恰是覆盖的执行者，而非防护）。
    companyBusinessSummary: baseline.companyBusinessSummary ?? null,
    meta: projected.meta
      ? { ...projected.meta, primaryInputMode: baseline.meta?.primaryInputMode ?? null }
      : baseline.meta ?? null,
    intakeItems: mergeByKeys(projected.intakeItems, baseline.intakeItems, projectedIntakeItemKeys),
    sectionTitles: mergeByKeys(projected.sectionTitles, baseline.sectionTitles, projectedSectionKeys),
    generatedBlocks: mergeByKeys(projected.generatedBlocks, baseline.generatedBlocks, projectedBlockIds),
    tableBlocks: mergeByKeys(projected.tableBlocks, baseline.tableBlocks, projectedBlockIds),
    imageBlocks: mergeByKeys(projected.imageBlocks, baseline.imageBlocks, projectedBlockIds),
  };
}

/** 生成用于 PUT 的完整 state:前端拥有的部分来自当前投影,投影之外的部分来自权威基线。 */
export function buildStoredReportStatePayload(
  report: Report,
  baseline: StoredReportStateV4 | null,
): StoredReportStateV4 {
  const projected = reportToStoredReportState(report);
  if (!baseline) return projected;
  const blockIds = new Set<string>();
  const sectionKeys = new Set<string>();
  const intakeItemKeys = new Set<string>();
  walkSections(report.sections ?? [], (section) => sectionKeys.add(section.key));
  walkBlocks(report.sections ?? [], (block) => blockIds.add(block.id));
  for (const item of report.intakeItems ?? []) intakeItemKeys.add(item.key);
  return mergeStoredStatePreservingUnprojected(projected, baseline, blockIds, sectionKeys, intakeItemKeys);
}

export function serializeStoredReportState(
  report: Report,
  baseline: StoredReportStateV4 | null = null,
): string {
  return JSON.stringify(buildStoredReportStatePayload(report, baseline));
}
